README


This archive contains the distribution Catmandu-RDF,
version 0.13:

  Modules for handling RDF data within the Catmandu framework

This software is copyright (c) 2014 by Jakob Voß.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




CHANGES
Changelog for Catmandu-RDF

0.13 2014-01-15T12:16:03
 - Merge release [Jakob Voss]
 - Merge branch 'master' of github.com:gbv/Catmandu-RDF [Jakob Voss]
 - use RDF::aREF 0.08 [Jakob Voss]

0.12 2014-01-12T16:39:03
 - release 0.12 with fixed dependencies [Jakob Voss]
 - update dependencies [Jakob Voss]
 - fixed dist.ini [Jakob Voss]
 - fixed dependency [Jakob Voss]

0.11 2013-12-02T14:17:01
 - documentation [Jakob Voss]
 - fixed cpanfile [Jakob Voss]
 - don't test ob Perl 5.10 [Jakob Voss]
 - split RDF::aREF as independent module [Jakob Voss]

0.10 2013-11-15T11:18:20
 - started to support nested input documents [Jakob Voss]
 - renamed aRef to aREF [Jakob Voss]

0.09 2013-11-13T14:29:24
 - migrated from RDF/JSON to aRef [Jakob Voss]
 - fix link [Jakob Voss]

0.03 2013-09-20T12:42:10
 - auto-expand prefixed object-uris [Jakob Voss]
 - support 'a' as shortcut for 'rdf:type' [Jakob Voss]
 - extended documentation [Jakob Voss]
 - implemented basic Exporter [Jakob Voss]
 - support a subset of JSON-LD [Jakob Voss]
 - use dzil and travis [Jakob Voss]

0.02 2013-09-06T09:14:51
 - documentation [Jakob Voss]

0.01 2013-09-05T09:39:00
 - basic RDF Exporter [Jakob Voss]
 - first draft [Jakob Voss]



lib/RDF/aREF.pm
package RDF::aREF;
#ABSTRACT: Another RDF Encoding Form
our $VERSION = '0.01';

use RDF::NS;

use parent 'Exporter';
#our @EXPORT = qw();

sub new {
    my ($class, %options) = @_;

    bless {
        ns => $options{ns} || RDF::NS->new()
    }, $class;
}
    #use Data::Dumper; say STDERR Dumper($rdf);


# requires Perl 5.12
use v5.12;
use feature 'unicode_strings';

our $nameChar = 'A-Z_a-z\N{U+00C0}-\N{U+00D6}\N{U+00D8}-\N{U+00F6}\N{U+00F8}-\N{U+02FF}\N{U+0370}-\N{U+037D}\N{U+037F}-\N{U+1FFF}\N{U+200C}-\N{U+200D}\N{U+2070}-\N{U+218F}\N{U+2C00}-\N{U+2FEF}\N{U+3001}-\N{U+D7FF}\N{U+F900}-\N{U+FDCF}\N{U+FDF0}-\N{U+FFFD}\N{U+10000}-\N{U+EFFFF}';
our $nameStartChar = $nameChar.'0-9\N{U+00B7}\N{U+0300}\N{U+036F}\N{U+203F}-\N{U+2040}-';
our $prefix = '[a-z]([a-z]|[0-9])*';
our $name   = "[$nameStartChar][$nameChar]*";
our $prefixedName = "$prefix:$name";

# TODO
our $plainIRI = qr{^[a-z][a-z0-9+.-]*:}i;

our $blankNode = qr{^_:([a-z0-9]+)$}i;

# object string
sub object_to_rdfjson {
    my ($self, $string) = @_;

    # absolute IRI enclosed in angle brackets
    if ($string =~ /^<(.+)>$/) {
        my $iri = $1; # TODO: validate IRI
        return { value => $iri, type => 'uri' };

    # blank node
    } elsif ($string =~ $blankNode) {
        return { value => $string, type => "bnode" };

    # prefixedName
    } elsif ($string =~ /^($prefix):($name)$/) {
        return { value => $self->prefixedName($1,$3), type => 'uri' };

    # languageTaggedString
    } elsif ($string =~ s/\@([a-z]{2,8}(-[a-z0-9]{1,8})*)$//) {
        return { value => $string, lang => $1, type => 'literal' };

    # datatypedString
    } elsif ($string =~ /^(.*)\^(($prefix):($name)|<(.+)>)$/) {

        my $value    = $1; 
        my $datatype;

        if (defined $3) {
            $datatype = $self->prefixedName($3,$5);
        } elsif (defined $6) {
            $datatype = $6; # TODO: validate IRI
        }
        
        $value =~ s/\^$//; 

        if ($datatype eq 'http://www.w3.org/2001/XMLSchema#string') {
            return { value => $value, type => 'literal' };
        } else {
            return { value => $value, type => 'literal', datatype => $datatype };
        }

    # explicitString
    } elsif ($string =~ s/\@$//) {
        return { value => $string, type => 'literal' };

    # plainIRI
    } elsif ($string =~ $plainIRI) {
        # TODO: validate syntax according to RFC 3987
        return { value => $string, type => 'uri' };
    }

    # simpleString
    return { value => $string, type => 'literal' };
}

sub prefixedName {
    my ($self, $prefix, $name) = @_;

    my $ns = $self->{ns}->{$prefix}
        or die "unknown prefix in $prefix:$name\n";

    # no IRI validation required with sane prefix definition
    return "$ns$name";
}

sub subject {
    my ($self, $string) = @_;

    my $subject = $string;

    # absolute IRI enclosed in angle brackets
    if ($subject =~ /^<(.+)>$/) {
        $subject = $1;

    # blank node
    } elsif ($subject =~ /^_:([a-zA-Z0-9]+)$/) {
        # $subject = $subject;

    # prefixed name (also with '_')
    } elsif ($subject =~ /^($prefix)[:_]($name)$/) {
        $subject = $self->prefixedName($1,$3);
    } 

    # TODO: should match plainIRI
    
    return $subject;
}

# TODO: implement to_iterator instead
sub to_rdfjson {
    my ($self, $graph) = @_;

    # TODO: _ns

    # property map
    if ($graph->{_id}) {
        my $subject = $self->subject($graph->{_id});    
        return $self->property_map_to_rdfjson( $subject => $graph ),

    # subject map
    } else {
        my $rdfjson = { };
        foreach my $subject ( grep { $_ ne '_ns' } keys %$graph ) {
            my $rdf = $self->property_map_to_rdfjson( 
                $self->subject($subject) => $graph->{$subject}
            );
            # merge (TODO: just return triples)
            foreach (keys %$rdf) {
                $rdfjson->{$_} = $rdf->{$_};
            }
        }
        return $rdfjson;
    }
}

sub property {
    my ($self, $string) = @_;

    # special shortcut 'a'
    if ($string eq 'a') {
        return 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type';

    # absolute IRI enclosed in angle brackets
    } elsif ($string =~ /^<(.+)>$/) {
        $string = $1;

    # prefixed name (also with '_')
    } elsif ($string =~ /^($prefix)[:_]($name)$/) {
        return $self->prefixedName($1,$3);
    }
    
    # TODO: validate IRI
    return $string;
}

sub property_map_to_rdfjson {
    my ($self, $subject, $map) = @_;
    
    my $statements = { };

    my $predicate_map = {
        map {
            my ($object, $stms) = 
                $self->encoded_object_to_rdfjson($map->{$_});

            # TODO: merge instead of replace, also different oid forms
            $statements = $stms if $stms and %$stms;

            ($self->property($_) => $object);
        } grep { $_ ne '_id' and $_ ne '_ns' } keys %$map
    };

    my $rdfjson = %$predicate_map ? { $subject => $predicate_map } : { };

    # Merge
    if (%$statements) {
        # TODO: merge instead replace
        for (keys %$statements) {
            $rdfjson->{$_} = $statements->{$_};
        }
    }
    
    return $rdfjson;
}

use Scalar::Util qw(reftype);

sub encoded_object_to_rdfjson {
    my ($self, $object) = @_;

    my $ref = reftype $object;

    if (!$ref) {
        return [ $self->object_to_rdfjson($object) ];
    } elsif($ref eq 'ARRAY') {
        return [ map { $self->object_to_rdfjson($_); } @$object ];
    } elsif($ref eq 'HASH') {
        my $id = $object->{_id};

        if (!defined $id) {
            $id = $self->blank_id;
        } elsif ($id =~ /^<(.+)>$/) {
            $id = $1;
        } elsif ($id =~ /^($prefix)[:_]($name)$/) {
            $id = $self->prefixedName($1,$3);
        }

        my $obj;
        if ($id =~ $blankNode) {
            $obj = { value => $id, type => "bnode" };
        } elsif ($id =~ $plainIRI) {
            $obj = { value => $id, type => "uri" };
        } else {
            die "expected IRI or blank node, got ".$object->{_id}."\n";
        }

        my $statements = $self->property_map_to_rdfjson( $id => $object );

        return [ $obj ], $statements;

    } else {
        print $object;
        die "encoded object must be a string, list, or predicate map!\n";
    }

}

sub blank_id {
    my ($self) = @_;
    return '_:'.++$self->{blank_counter};
}

1;


__END__
=pod

=head1 NAME

RDF::aREF - Another RDF Encoding Form

=head1 VERSION

version 0.10

=head1 SYNOPSIS

  my $aref = {
     # aREF RDF data
  };

  my $rdfjson = RDF::aREF->new->to_rdfjson( $aref );
  RDF::Trine::Model->add_hashref( $rdfjson );

=head1 DESCRIPTION

This module implements a parser of Another RDF encoding form (aREF). The module
is shipped with L<Catmandu::RDF> but will be refactored to be published as
independent module on CPAN. As aREF is not finally specified, this module is in
a very early state of development!

=head1 SEE ALSO

aREF is being specified at L<http://github.com/gbv/aref>.

See L<RDF::YAML> for an outdated parser/serializer of a similar RDF encoding in
YAML.

=encoding utf8

=head1 AUTHOR

Jakob Voß

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Jakob Voß.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/RDF/aRef.pm
package RDF::aRef;
#ABSTRACT: Another RDF Encoding Form
our $VERSION = '0.01';

use RDF::NS;

use parent 'Exporter';
#our @EXPORT = qw();

sub new {
    my ($class, %options) = @_;

    bless {
        ns => $options{ns} || RDF::NS->new()
    }, $class;
}
    #use Data::Dumper; say STDERR Dumper($rdf);


# requires Perl 5.12
use feature 'unicode_strings';

our $nameChar = 'A-Z_a-z\N{U+00C0}-\N{U+00D6}\N{U+00D8}-\N{U+00F6}\N{U+00F8}-\N{U+02FF}\N{U+0370}-\N{U+037D}\N{U+037F}-\N{U+1FFF}\N{U+200C}-\N{U+200D}\N{U+2070}-\N{U+218F}\N{U+2C00}-\N{U+2FEF}\N{U+3001}-\N{U+D7FF}\N{U+F900}-\N{U+FDCF}\N{U+FDF0}-\N{U+FFFD}\N{U+10000}-\N{U+EFFFF}';
our $nameStartChar = $nameChar.'0-9\N{U+00B7}\N{U+0300}\N{U+036F}\N{U+203F}-\N{U+2040}-';
our $prefix = '[a-z]([a-z]|[0-9])*';
our $name   = "[$nameStartChar][$nameChar]*";
our $prefixedName = "$prefix:$name";

# TODO
our $SimplfiedIRIRegexp = qr{^[a-z][a-z0-9+.-]*:}i;

# object string
sub object_to_rdfjson {
    my ($self, $string) = @_;

    # absolute IRI enclosed in angle brackets
    if ($string =~ /^<(.+)>$/) {
        my $iri = $1; # TODO: validate IRI
        return { value => $iri, type => 'uri' };

    # blank node
    } elsif ($string =~ /^_:([a-zA-Z0-9]+)$/) {
        return { value => $string, type => "bnode" };

    # prefixedName
    } elsif ($string =~ /^($prefix):($name)$/) {
        return { value => $self->prefixedName($1,$3), type => 'uri' };

    # languageTaggedString
    } elsif ($string =~ s/\@([a-z]{2,8}(-[a-z0-9]{1,8})*)$//) {
        return { value => $string, lang => $1, type => 'literal' };

    # datatypedString
    } elsif ($string =~ /^(.*)\^(($prefix):($name)|<(.+)>)$/) {

        my $value    = $1; 
        my $datatype;

        if (defined $3) {
            $datatype = $self->prefixedName($3,$5);
        } elsif (defined $6) {
            $datatype = $6; # TODO: validate IRI
        }
        
        $value =~ s/\^$//; 

        if ($datatype eq 'http://www.w3.org/2001/XMLSchema#string') {
            return { value => $value, type => 'literal' };
        } else {
            return { value => $value, type => 'literal', datatype => $datatype };
        }

    # explicitString
    } elsif ($string =~ s/\@$//) {
        return { value => $string, type => 'literal' };

    # plainIRI
    } elsif ($string =~ $SimplfiedIRIRegexp) {
        # TODO: validate more (?)
        return { value => $string, type => 'uri' };
    }

    # simpleString
    return { value => $string, type => 'literal' };
}

sub prefixedName {
    my ($self, $prefix, $name) = @_;

    my $ns = $self->{ns}->{$prefix}
        or die "unknown prefix in $prefix:$name\n";

    # no IRI validation required with sane prefix definition
    return "$ns$name";
}

sub subject {
    my ($self, $string) = @_;

    my $subject = $string;

    # absolute IRI enclosed in angle brackets
    if ($subject =~ /^<(.+)>$/) {
        $subject = $1;

    # blank node
    } elsif ($subject =~ /^_:([a-zA-Z0-9]+)$/) {
        # $subject = $subject;

    # prefixed name (also with '_')
    } elsif ($subject =~ /^($prefix)[:_]($name)$/) {
        $subject = $self->prefixedName($1,$3);
    }

    return $subject;
}

sub to_rdfjson {
    my ($self, $graph) = @_;

    # property map
    if ($graph->{_id}) {
        my $subject = $self->subject($graph->{_id});    

        return {
            $subject => $self->property_map_to_rdfjson($graph),
        };
    # subject map
    } else {
        return {
            map { 
                ($self->subject($_) =>  
                 $self->property_map_to_rdfjson($graph->{$_}))
            }
            keys %$graph
        };
        # TODO
    }
}

sub property {
    my ($self, $string) = @_;

    # special shortcut 'a'
    if ($string eq 'a') {
        return 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type';

    # absolute IRI enclosed in angle brackets
    } elsif ($string =~ /^<(.+)>$/) {
        $string = $1;

    # prefixed name (also with '_')
    } elsif ($string =~ /^($prefix)[:_]($name)$/) {
        return $self->prefixedName($1,$3);
    }
    
    # TODO: validate IRI
    return $string;
}

sub property_map_to_rdfjson {
    my ($self, $map) = @_;
    return { 
        map {
            $self->property($_) => $self->encoded_object_to_rdfjson($map->{$_})
        } grep { $_ ne '_id' and $_ ne '_ns' } keys %$map
    }
}

sub encoded_object_to_rdfjson {
    my ($self, $object) = @_;

    if (!ref $object) {
        return [ $self->object_to_rdfjson($object) ];
    } elsif(ref $object ne 'ARRAY') {
        die "encoded object must be a list!\n";
    }

    return [ map { $self->object_to_rdfjson($_); } @$object ];
}

1;


__END__
=pod

=head1 NAME

RDF::aRef - Another RDF Encoding Form

=head1 VERSION

version 0.09

=head1 SYNOPSIS

  my $aref = {
     # aRef RDF data
  };

  my $rdfjson = RDF::aRef->new->to_rdfjson( $aref );
  RDF::Trine::Model->add_hashref( $rdfjson );

=head1 DESCRIPTION

This module implements a parser of Another RDF encoding form (aRef). The module
is shipped with L<Catmandu::RDF> but will be refactored to be published as
independent module on CPAN. As aRef is not finally specified, this module is in
a very early state of development!

=head1 SEE ALSO

aRef is being specified at L<http://github.com/gbv/aref>.

See L<RDF::YAML> for an outdated parser/serializer of a similar RDF encoding in
YAML.

=head1 AUTHOR

Jakob Voß

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Jakob Voß.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Catmandu/RDF.pm
use strict;
use warnings;
package Catmandu::RDF;
# ABSTRACT: Modules for handling RDF data within the Catmandu framework
our $VERSION = '0.13'; # VERSION


1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Catmandu::RDF - Modules for handling RDF data within the Catmandu framework

=head1 VERSION

version 0.13

=head1 DESCRIPTION

Catmandu::RDF contains modules for handling RDF data within the L<Catmandu>
framework. This release is in an early state of development. Feedback and
contributions are very welcome at L<https://github.com/nichtich/Catmandu-RDF>!

=head1 AVAILABLE MODULES

=over 4

=item L<Catmandu::Exporter::RDF>

Serialize RDF data (as RDF/XML, RDF/JSON, Turtle, NTriples, RDFa...).
RDF data must be provided in B<another RDF Encoding Form (aREF)> as 
implemented with L<RDF::aREF>.

=back

=head1 SUGGESTED MODULES

The following modules have not been implemented yet. Please contribute or
comment if you miss them!

=over 4

=item C<Catmandu::Importer::RDF>

Import serialized RDF data (RDF/XML, RDF/JSON, Turtle, NTriples, RDFa...).

=item C<Catmandu::Importer::SPARQL>

Import RDF data from a SPARQL endpoint.

=item C<Catmandu::Exporter::SPARUL> or C<Catmandu::Exporter::SPARQL>

Export RDF data with SPARQL/Update.

=item C<Catmandu::Exporter::RDFPatch>

Export RDF with HTTP PATCH.

=back

=head1 SEE ALSO

L<Catmandu>, L<RDF::aREF>, L<RDF::Trine>, L<RDF::NS>

=head1 AUTHOR

Jakob Voß

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Jakob Voß.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Modules for handling RDF data within the Catmandu framework",
  "AUTHOR" => "Jakob Vo\x{df}",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Catmandu-RDF",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Catmandu::RDF",
  "PREREQ_PM" => {
    "Catmandu" => "0.7",
    "RDF::NS" => 20130816,
    "RDF::Trine" => "1.0",
    "RDF::aREF" => "0.08"
  },
  "TEST_REQUIRES" => {},
  "VERSION" => "0.13",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Catmandu" => "0.7",
  "RDF::NS" => 20130816,
  "RDF::Trine" => "1.0",
  "RDF::aREF" => "0.08"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



